### PR TITLE
Pad functions

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- New Feature

Changelog entry:
Add new functions `leftPad()`, `rightPad()`, `leftPadUTF8()`, `rightPadUTF8()`.

---
This PR is to finish the work started by @dankondr (https://github.com/ClickHouse/ClickHouse/pull/19575)

### CODE REVIEW

File: src/Functions/GatherUtils/Sources.h [hunk 1]
```diff
@@ -755,6 +755,7 @@ struct GenericValueSource : public ValueSourceImpl<GenericValueSource>
 {
     using Slice = GenericValueSlice;
     using SinkType = GenericArraySink;
+    using Column = IColumn;
```
Comment (kitaisreal) [inline src/Functions/GatherUtils/Sources.h:758]: @vitlibar why is that needed?

File: src/Functions/padString.cpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+                    if (offset == pad_string.length())
+                        break;
+                    offset += UTF8::seqLength(pad_string[offset]);
+                    if (offset > pad_string.length())
+                        offset = pad_string.length();
+                }
+            }
+
+            /// Not necessary, but good for performance.
+            while (numCharsInPadString() < 16)
```
Comment (kitaisreal) [inline src/Functions/padString.cpp:92]: @vitlibar can you add some comment what is going on here ?

File: src/Functions/padString.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                writeSlice(StringSource::Slice{bit_cast<const UInt8 *>(pad_string.data()), numCharsToNumBytes(step)}, res_sink);
+                num_chars -= step;
+            }
+        }
+
+    private:
+        void init()
+        {
+            if (pad_string.empty())
+                pad_string = " ";
```
Comment (kitaisreal) [inline src/Functions/padString.cpp:74]: Consider to remove temporary string creation pad_string = ' ';

File: src/Functions/padString.cpp [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        void executeForSourceAndLength(
+            SourceStrings && strings,
+            SourceLengths && lengths,
+            const PaddingChars<is_utf8> & padding_chars,
+            StringSink & res_sink) const
+        {
+            bool is_const_length = lengths.isConst();
+            bool need_check_length = true;
+
+            for (; !res_sink.isEnd(); res_sink.next(), strings.next(), lengths.next())
```
Comment (kitaisreal) [inline src/Functions/padString.cpp:249]: @vitlibar add some comment here.

File: src/Functions/padString.cpp (hunk 3 above)
Comment (vitlibar) [reply src/Functions/padString.cpp:74]: I don't think `pad_string = ' '` is better than `pad_string = " "` here. I mean because of we're assigning a string (and not a char) here, the syntax with double quotas is clearer. If we compare performance then I agree `pad_string = ' '` can be a little bit faster, but that difference is negligibly small.

File: src/Functions/GatherUtils/Sources.h (hunk 1 above)
Comment (vitlibar) [reply src/Functions/GatherUtils/Sources.h:758]: It wouldn't be compiled without this line, because I'm using the type [ConstSource<GenericValueSource>](https://github.com/ClickHouse/ClickHouse/pull/26075/files#diff-c80e123b3b061659640ed20e8cbe00a51b326db4493e57bf2b405c0650790a38R234) and the template class `ConstSource` depends on the type [Base::Column](https://github.com/vitlibar/ClickHouse/blob/3dfdcf4604e4c13e3b90743cede89cc5ed942b5c/src/Functions/GatherUtils/Sources.h#L168).

File: src/Functions/padString.cpp (hunk 2 above)
Comment (vitlibar) [reply src/Functions/padString.cpp:92]: https://github.com/ClickHouse/ClickHouse/pull/26147

File: src/Functions/padString.cpp (hunk 4 above)
Comment (vitlibar) [reply src/Functions/padString.cpp:249]: https://github.com/ClickHouse/ClickHouse/pull/26147

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11556.
